### DISCUSSION TITLE
v4 Roadmap

### DISCUSSION THREAD

**Discussion by JimBobSquarePants** [SixLabors/ImageSharp#2533] [Ideas]
WIP. General Roadmap for v4. 

- [x] Target .NET 8
- [x] #2531 
- [x] #2532 
- [x] #2534 
- [x] Revisit SIMD pixel conversion pipelines and simplify/optimize (Are we overusing Unsafe API?).
- [ ] Performance pass for RGBA compatible converters.
- [ ] Expose general Convolution API
- [ ] #1567
- [x] #1982
- [x] [How to make libraries compatible with native AOT](https://devblogs.microsoft.com/dotnet/creating-aot-compatible-libraries) This could possibly be a shared configuration.
- [x] Add general metadata formats to allow easy conversion between explicit formats.
- [ ] Enhance color cache to allow greater precision for Non-alpha pixel fomats.
- [x] Add API for metadata conversion between formats.

**Comment by stefannikolei** [comment]
Do you see any use in new C#12 Features e.g. Inline Arrays?

**Reply by JimBobSquarePants** [reply]
Most likely yes. We just need to identify opportunities.

**Comment by sgf** [comment]
Is ImageSharp a vector graphics library? What level does it target? For example, can it replace Skia, Cairo, impeller, Nanovg and other libraries? From the name, it seems to be an image processing library. Does this name affect the spread of the library? Have you considered extracting the core part, creating a new graphics library, and connecting to multiple hardware accelerated backends (gl, gles, vulkan, metal, dx...)? Of course, a short and catchy name is also very important.

Microsoft has no plans to create a cross-platform 2D graphics library. This is actually an opportunity for the .NET community. If Microsoft can be persuaded to provide appropriate financial support, perhaps this will produce a good open source project. Of course, the complexity of the 2D graphics library is very high, and this will be a big project.

Impeller may be a very valuable library for reference. Because it is used as Flutter's 2D graphics infrastructure. Moreover, it is newly developed, so the architecture is more modern.
